File: packages/contracts/v1/event-stream.ts
import type { UUID, ISO8601 } from "./common.js";

export type JobType = string;

export interface JobUpdateEvent {
  event: "job_update";
  data: {
    job_id: UUID;
    job_type: JobType;
    status: "queued" | "running" | "complete" | "failed" | "retrying";
    retry_count: number;
    progress: number;
    stage: string | null;
    result_ref: UUID | null;
    error: { code: string; message: string } | null;
    timestamp: ISO8601;
  };
}

export interface ProjectStatusEvent {
  event: "project_status";
  data: Record<string, unknown>;
}

export interface StageProgressEvent {
  event: "stage_progress";
  data: Record<string, unknown>;
}

export interface AmbiguityFlagEvent {
  event: "ambiguity_flag";
  data: {
    flag_id: UUID;
    element_type: string;
    description: string;
    confidence: number;
    requires_user_action: boolean;
    timestamp: ISO8601;
  };
}

export interface CritiqueSummaryEvent {
  event: "critique_summary";
  data: Record<string, unknown>;
}

export interface HeartbeatEvent {
  event: "heartbeat";
  data: { timestamp: ISO8601 };
}

export type SSEEvent =
  | JobUpdateEvent
  | ProjectStatusEvent
  | StageProgressEvent
  | AmbiguityFlagEvent
  | CritiqueSummaryEvent
  | HeartbeatEvent;

File: packages/contracts/python/noa_contracts/event_stream.py
from __future__ import annotations

from typing import Literal
from pydantic import BaseModel


class JobUpdateData(BaseModel):
    job_id: str
    job_type: str
    status: Literal["queued", "running", "complete", "failed", "retrying"]
    retry_count: int
    progress: float
    stage: str | None
    result_ref: str | None
    error: dict[str, str] | None
    timestamp: str


class JobUpdateEvent(BaseModel):
    event: Literal["job_update"] = "job_update"
    data: JobUpdateData


class AmbiguityFlagData(BaseModel):
    flag_id: str
    element_type: str
    description: str
    confidence: float
    requires_user_action: bool
    timestamp: str


class AmbiguityFlagEvent(BaseModel):
    event: Literal["ambiguity_flag"] = "ambiguity_flag"
    data: AmbiguityFlagData


class HeartbeatEvent(BaseModel):
    event: Literal["heartbeat"] = "heartbeat"
    data: dict[str, str]  # timestamp
